### PR TITLE
Add end to end test for MR async job

### PR DESCRIPTION
**Pull request by lugi0**
When the S3/OCI fixtures are ready, we can simply:
  - Replace placeholder_s3_secret_name and placeholder_oci_secret_name fixtures in conftest.py
  - Update the placeholder constants in constants.py
  - Add end-to-end execution tests

## Summary by CodeRabbit

* **Tests**
  * Added an end-to-end test validating asynchronous model upload (object storage → OCI registry → Model Registry) and skipped one OCI registry test slated for removal.
* **Fixtures**
  * Added fixtures for S3/OCI credentials, async job deployment with volume mounts, OCI host exposure, preloading sample models (including from images), and registered-model provisioning.
* **Utilities**
  * Added helpers to find job pods, encode S3 secrets, upload artifacts to object storage from images, and push/pull OCI manifests.
* **Documentation / Config**
  * Centralized async job configuration into shared constants.
* **Chores**
  * Reworked RBAC/test access: introduced service-account token flows and consolidated group-based access fixtures, streamlining test auth setup.

### CODE REVIEW

File: tests/model_registry/async_job/utils.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            label_selector=f"job-name={job.name}",
+        )
+    )
+    assert len(pods) == 1, f"Expected 1 pod for job {job.name}, found {len(pods)}"
+    return pods[0]
+
+
+def validate_job_labels_and_annotations(job: Job) -> None:
+    """Validate Job has all required labels and annotations"""
+    job_labels = job.instance.metadata.labels or {}
```
Comment (dbasunag) [inline tests/model_registry/async_job/utils.py:28]: Save `job.instance.metadata`, that way you don't need to do two api calls for labels and annotations.

File: tests/model_registry/async_job/test_model_sync_job.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    )
+    def test_job_pod_creation(self: Self, admin_client: DynamicClient, model_sync_async_job: Job) -> None:
+        """Verify Job creates a pod (when image is available)"""
+        job_pod = get_job_pod(admin_client=admin_client, job=model_sync_async_job)
+        assert job_pod.name.startswith(ASYNC_UPLOAD_JOB_NAME)
+
+    def test_job_model_sync_label_propagation(self: Self, model_sync_async_job: Job) -> None:
+        """Verify model sync labels are properly set"""
+        # Test that model sync labels are present on Job
+        job_labels = model_sync_async_job.instance.metadata.labels
```
Comment (dbasunag) [inline tests/model_registry/async_job/test_model_sync_job.py:85]: Should we wait for the job to be completed first?

File: tests/model_registry/async_job/utils.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+    # Validate required annotations
+    for key, expected_value in ASYNC_JOB_ANNOTATIONS.items():
+        actual_value = job_annotations.get(key)
+        assert actual_value == expected_value, f"Annotation {key}: expected {expected_value}, got {actual_value}"
+
+
+def validate_job_pod_template_labels(job: Job) -> None:
+    """Validate Job pod template has required model sync labels"""
+    pod_template_labels = job.instance.spec.template.metadata.labels or {}
```
Comment (dbasunag) [inline tests/model_registry/async_job/utils.py:44]: All the validations would require, api calls. Do you want to send the job dict instead of job object?

Comment (lugi0) [reply tests/model_registry/async_job/utils.py:44]: Wouldn't this require a single api call here and then the labels are stored in pod_template_labels, and we're just fetching them by key a few lines down?

File: tests/model_registry/async_job/constants.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Job identification
+ASYNC_UPLOAD_JOB_NAME = "model-sync-async-job"
+
+# Midstream temporary image (until downstream is available)
+# TODO: Update before async job goes live
+ASYNC_UPLOAD_IMAGE = "quay.io/opendatahub/model-registry-job-async-upload:main-500be9d"
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/constants.py:6]: Not sure how we will handle it once live, but we shouldn't keep using the midstream image

**Comment by lugi0** [comment]
@mwaykole only way I can think of to make this work in disconnected environments

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+        )
+
+        # Verify model was created
+        assert registered_model.id, "Model should have an ID"
+        assert registered_model.name == test_model_data["model_name"]
+
+        LOGGER.info(f"Created test model: {registered_model.name} (ID: {registered_model.id})")
+        LOGGER.info("Model registry client integration: ✅ PASSED")
+
+    def test_infrastructure_connectivity(
```
Comment (fege) [inline tests/model_registry/async_job/test_async_upload_e2e.py:96]: I think this can be removed or integrated in `test_end_to_end_async_upload_job`

File: tests/model_registry/async_job/constants.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+MODEL_SYNC_CONFIG = {
+    "MODEL_ID": "1",
+    "MODEL_VERSION_ID": "2",
+    "MODEL_ARTIFACT_ID": "1",
+    "SOURCE_TYPE": "s3",
+    "DESTINATION_TYPE": "oci",
+    "SOURCE_AWS_KEY": "my-model",
+    "DESTINATION_OCI_BASE_IMAGE": "public.ecr.aws/docker/library/busybox:latest",
+    "DESTINATION_OCI_ENABLE_TLS_VERIFY": "false",
+}
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/constants.py:31]: not needed, we start from a clean known state each time

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+            LOGGER.info("✓ Manifest found in OCI registry")
+            LOGGER.info(f"✓ Manifest schema version: {manifest.get('schemaVersion')}")
+            LOGGER.info(f"✓ Manifest media type: {manifest.get('mediaType')}")
+
+            # Verify the manifest has the expected structure
+            # LOGGER.info(f"✓ Manifest: {manifest}")
+            assert "manifests" in manifest, "Manifest should contain manifests section"
+            assert len(manifest["manifests"]) > 0, "Manifest should have at least one manifest"
+
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/test_async_upload_e2e.py:356]: this has been verified and it is the correct format

File: tests/model_registry/async_job/utils.py [hunk 8]
```diff
@@ -1,12 +1,37 @@
...
+
+    if not pods:
+        raise AssertionError(f"No pods found for job {job.name}")
+
+    # Sort pods by creation time (latest first)
+    sorted_pods = sorted(pods, key=lambda p: p.instance.metadata.creationTimestamp or "", reverse=True)
+
+    latest_pod = sorted_pods[0]
+    LOGGER.info(f"Found {len(pods)} pod(s) for job {job.name}, using latest: {latest_pod.name}")
+    return latest_pod
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/utils.py:32]: not needed, it's an edge case

File: tests/model_registry/async_job/utils.py [hunk 9]
```diff
@@ -71,3 +96,130 @@ def pull_manifest_from_oci_registry(registry_url: str, repo: str, tag: str) -> d
...
+                    "capabilities": {"drop": ["ALL"]},
+                    "runAsNonRoot": True,
+                    "seccompProfile": {"type": "RuntimeDefault"},
+                },
+            }
+        ],
+        wait_for_resource=True,
+    ) as upload_pod:
+        LOGGER.info(f"Uploading model file to MinIO: {object_key}")
+        upload_pod.wait_for_status(status="Succeeded", timeout=300)
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/utils.py:216]: it works fine for now and the model is small enough that this is not a concern

File: tests/model_registry/async_job/test_async_upload_e2e.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/test_async_upload_e2e.py:356]: it does not fail, this is the format that is being used

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 10]
```diff
@@ -0,0 +1,381 @@
...
+                        LOGGER.warning(f"⚠ Pod {current_pod.name} failed, job may retry with new pod")
+                        # Don't fail here - let the job retry
+                    elif pod_status.phase == "Running":
+                        LOGGER.info(f"Pod {current_pod.name} is running...")
+
+            except Exception as e:
+                LOGGER.warning(f"Could not get pod status: {e}")
+
+            LOGGER.info(f"Job still running... (elapsed: {int(time.time() - start_time)}s)")
+            time.sleep(seconds=10)
```
[the remark this answers is not in the document]
Comment (lugi0) [reply tests/model_registry/async_job/test_async_upload_e2e.py:321]: it has the seconds keyword arg, this is fine.

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+    indirect=True,
+)
+class TestAsyncUploadE2E:
+    """Test for async upload job with real MinIO, OCI registry, and Model Registry"""
+
+    @pytest.mark.skipif(
+        ASYNC_UPLOAD_IMAGE.startswith("PLACEHOLDER"),
+        reason="Downstream image not yet available - job will fail to start",
+    )
+    def test_async_upload_job(
```
Comment (adolfo-ab) [inline tests/model_registry/async_job/test_async_upload_e2e.py:72]: This is up to you, but would it make sense to split this test into smaller ones within the same test class?

File: tests/model_registry/async_job/utils.py [hunk 12]
```diff
@@ -1,12 +1,75 @@
...
+    Args:
+        admin_client: Kubernetes dynamic client
+        job: Job object to check
+
+    Returns:
+        bool: True if job completed successfully, False otherwise
+    """
+    job_status = job.instance.status
+
+    # Check if job succeeded
```
Comment (adolfo-ab) [inline tests/model_registry/async_job/utils.py:28]: this comment doesn't add much

File: tests/model_registry/async_job/utils.py [hunk 13]
```diff
@@ -71,3 +134,129 @@ def pull_manifest_from_oci_registry(registry_url: str, repo: str, tag: str) -> d
...
+
+    Args:
+        admin_client: Kubernetes client
+        namespace: Namespace to create upload pod in
+        minio_service: MinIO service resource
+        object_key: S3 object key path (default: "my-model/mnist-8.onnx")
+    """
+    from utilities.constants import MinIo
+    import base64
+    import os
```
Comment (adolfo-ab) [inline tests/model_registry/async_job/utils.py:186]: We usually put all the imports at the top of the files

File: tests/model_registry/conftest.py [hunk 14]
```diff
@@ -296,3 +306,153 @@ def model_registry_pod(admin_client: DynamicClient, model_registry_namespace: st
     )
     assert len(mr_pod) == 1
     return mr_pod[0]
+
+
+@pytest.fixture(scope="function")
+def sa_namespace(request: pytest.FixtureRequest, admin_client: DynamicClient) -> Generator[Namespace, None, None]:
+    """
+    Creates a temporary namespace using a context manager for automatic cleanup.
+    Function scope ensures a fresh namespace for each test needing it.
```
Comment (adolfo-ab) [inline tests/model_registry/conftest.py:315]: suspicious comment :robot: :eyes:

File: tests/model_registry/conftest.py [hunk 15]
```diff
@@ -296,3 +306,153 @@ def model_registry_pod(admin_client: DynamicClient, model_registry_namespace: st
...
+    with Namespace(client=admin_client, name=ns_name) as ns:
+        ns.wait_for_status(status=Namespace.Status.ACTIVE, timeout=120)
+        yield ns
+
+
+@pytest.fixture(scope="function")
+def service_account(admin_client: DynamicClient, sa_namespace: Namespace) -> Generator[ServiceAccount, None, None]:
+    """
+    Creates a ServiceAccount within the temporary namespace using a context manager.
+    Function scope ensures it's tied to the lifetime of sa_namespace for that test.
```
Comment (adolfo-ab) [inline tests/model_registry/conftest.py:329]: same

File: tests/model_registry/conftest.py [hunk 16]
```diff
@@ -296,3 +306,153 @@ def model_registry_pod(admin_client: DynamicClient, model_registry_namespace: st
...
+            "resourceNames": [MR_INSTANCE_NAME],  # Grant access only to the specific MR service object
+            "verbs": ["get"],
+        }
+    ]
+    role_labels = {
+        "app.kubernetes.io/component": "model-registry-test-rbac",
+        "test.opendatahub.io/namespace": sa_namespace.name,
+    }
+
+    LOGGER.info(f"Attempting to create Role: {role_name} with rules and labels.")
```
Comment (adolfo-ab) [inline tests/model_registry/conftest.py:407]: I think this doesn't add much, the wrapper will already log the creation request sent to the API

File: tests/model_registry/conftest.py [hunk 17]
```diff
@@ -296,3 +306,153 @@ def model_registry_pod(admin_client: DynamicClient, model_registry_namespace: st
...
+    with Role(
+        client=admin_client,
+        name=role_name,
+        namespace=model_registry_namespace,
+        rules=role_rules,
+        label=role_labels,
+        wait_for_resource=True,
+    ) as role:
+        LOGGER.info(f"Role {role.name} created successfully.")
```
Comment (adolfo-ab) [inline tests/model_registry/conftest.py:416]: same

File: tests/model_registry/conftest.py [hunk 18]
```diff
@@ -296,3 +306,153 @@ def model_registry_pod(admin_client: DynamicClient, model_registry_namespace: st
...
+    LOGGER.info(
+        f"Defining RoleBinding: {binding_name} linking Group 'system:serviceaccounts:{sa_namespace.name}' "
+        f"to Role '{mr_access_role.name}' in namespace {model_registry_namespace}"
+    )
+    binding_labels = {
+        "app.kubernetes.io/component": "model-registry-test-rbac",
+        "test.opendatahub.io/namespace": sa_namespace.name,
+    }
+
+    LOGGER.info(f"Attempting to create RoleBinding: {binding_name} with labels.")
```
Comment (adolfo-ab) [inline tests/model_registry/conftest.py:441]: same

File: tests/model_registry/async_job/utils.py [hunk 19]
```diff
@@ -71,3 +133,130 @@
...
+                "args": [
+                    # Set up MinIO client and upload the model file
+                    f"export MC_CONFIG_DIR=/upload-data/.mc && "
+                    f"mc alias set testminio http://{minio_service.name}.{minio_service.namespace}.svc.cluster.local:{MinIo.Metadata.DEFAULT_PORT} "  # noqa: E501
+                    f"{MinIo.Credentials.ACCESS_KEY_VALUE} {MinIo.Credentials.SECRET_KEY_VALUE} && "
+                    f"mc mb --ignore-existing testminio/{MinIo.Buckets.MODELMESH_EXAMPLE_MODELS} && "
+                    f"mc cp /upload-data/model-file testminio/{MinIo.Buckets.MODELMESH_EXAMPLE_MODELS}/{object_key} && "
+                    f"mc ls testminio/{MinIo.Buckets.MODELMESH_EXAMPLE_MODELS}/my-model/ && "
+                    f"echo 'Upload completed successfully'"
+                ],
```
[the remark this answers is not in the document]
Comment (fege) [reply tests/model_registry/async_job/utils.py]: no needed

File: tests/model_registry/async_job/test_async_upload_e2e.py (hunk 11 above)
Comment (fege) [reply tests/model_registry/async_job/test_async_upload_e2e.py:72]: I prefer to keep a single test for the e2e flow

File: tests/model_registry/async_job/utils.py [hunk 20]
```diff
@@ -1,12 +1,75 @@
...
+            if pod_status.phase == "Succeeded":
+                LOGGER.info(f"Pod {current_pod.name} succeeded")
+            elif pod_status.phase == "Failed":
+                LOGGER.warning(f"Pod {current_pod.name} failed, job may retry with new pod")
+            elif pod_status.phase == "Running":
+                LOGGER.info(f"Pod {current_pod.name} is running...")
+            else:
+                LOGGER.info(f"Pod {current_pod.name} in phase: {pod_status.phase}")
+    except Exception as e:
+        LOGGER.warning(f"Could not get pod status: {e}")
```
Comment (dbasunag) [inline tests/model_registry/async_job/utils.py:47]: These are log only. So they would have no impact on your run.

Comment (fege) [reply tests/model_registry/async_job/utils.py:47]: yes, thanks for noticing

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+        # Wait for job to complete
+        try:
+            for sample in TimeoutSampler(
+                wait_timeout=300,  # 5 minutes
+                sleep=10,  # Check every 10 seconds
+                func=lambda: check_job_completion(admin_client=admin_client, job=model_sync_async_job),
+            ):
+                if sample:  # Job completed successfully
+                    break
```
[the remark this answers is not in the document]
Comment (fege) [reply tests/model_registry/async_job/test_async_upload_e2e.py:97]: you are right, changing it

File: tests/model_registry/async_job/test_async_upload_e2e.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+        reason="Downstream image not yet available - job will fail to start",
+    )
+    def test_async_upload_job(
+        self: Self,
+        admin_client: DynamicClient,
+        model_sync_async_job: Job,
+        model_registry_client: list[ModelRegistryClient],
+        oci_registry_route: Route,
+    ) -> None:
+        """Test for async upload job execution using model from KSERVE_MINIO_IMAGE"""
```
[the remark this answers is not in the document]
Comment (fege) [reply tests/model_registry/async_job/test_async_upload_e2e.py:79]: can we do in a cleanup pr? there are also the older test/example to remove

[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/model_registry/async_job/test_async_upload_e2e.py:72]: Sure. I will resolve it for now.

**Review by dbasunag — APPROVED** [review]

**Review by sheltoncyril — APPROVED** [review]
/lgtm
